apps/workspace-host/src/interface.js

```diff
@@ -210,16 +231,24 @@ async
           } catch (err) {
             debug(`Couldn't find container: ${err}`);
           }
-        } else if (ws.state === 'running') {
-          if (!ws.launch_uuid) {
-            await workspaceUtils.updateWorkspaceState(ws.id, 'stopped', 'Shutting down');
-            await sqldb.queryAsync(sql.clear_workspace_on_shutdown, {
-              workspace_id: ws.id,
-              instance_id: workspace_server_settings.instance_id,
-            });
-          }
         }
       });
+
+      // Especially in dev mode, there may be containers that are running but
+      // don't correspond to any known workspace. Clean them up.
+      const allContainers = await docker.listContainers({
+        all: true,
+        filters: { label: ['prairielearn.workspace-id'] },
+      });
+      for (const container of allContainers) {
+        const containerWorkspaceId = container.Labels['prairielearn.workspace-id'];
```

Comment:
For dev mode outside of a container we could be even more polite and leave any container without a `prairielearn.workspace-id` label running. These would presumably be other containers running on the dev box, unrelated to PL. I _think_ this would still do what we want in prod, because we presumably label all of our containers?

Reply:
I think that's what line 241 is doing?

Reply:
Yeah, that was the intention of that line. Per https://docs.docker.com/engine/reference/commandline/ps, `label=foo` (which is what that line should translate to) finds only images where the label `foo` is present.